editoast/src/views/pathfinding/catenaries.rs

```diff
@@ -0,0 +1,557 @@
+    clip_range: Range<Float>,
+) -> RangeMap<Float, String> {
+    let mut res = RangeMap::new();
+    for (range, value) in range_map.overlapping(&clip_range) {
+        let range = range.start.max(clip_range.start)..range.end.min(clip_range.end);
+        res.insert(range, value.clone());
+    }
+    res
+}
+
+/// Extend the pathfinding range map with the other one, as though entering from `range_entry` in the
+/// given direction.
+fn extend_path_range_map(
+    path_range_map: &mut RangeMap<Float, String>,
+    other_range_map: &RangeMap<Float, String>,
+    range_entry: f64,
+    offset: f64,
+    direction: Direction,
+) {
+    for (range, value) in other_range_map.iter() {
+        let (start, end) = match direction {
+            Direction::StartToStop => (range.start.0, range.end.0),
+            Direction::StopToStart => (range.end.0, range.start.0),
+        };
+
+        let range = ((start - range_entry).abs() + offset).into()
+            ..((end - range_entry).abs() + offset).into();
+        assert!(!path_range_map.overlaps(&range), "range overlap");
+        path_range_map.insert(range, value.clone());
+    }
+}
+
+fn make_path_range_map(
+    catenary_mode_map: &HashMap<String, RangeMap<Float, String>>,
+    pathfinding: &Pathfinding,
+) -> RangeMap<Float, String> {
+    let mut res = RangeMap::new();
+    let mut offset = 0.;
+
+    for track_range in pathfinding.merged_track_ranges() {
+        let mode_range_map_option = catenary_mode_map.get(&track_range.track as &str);
+        if let Some(mode_range_map) = mode_range_map_option {
+            let mode_range_map = clip_range_map(
+                mode_range_map,
+                track_range.begin.into()..track_range.end.into(),
+            );
+            let range_entry = match track_range.direction {
+                Direction::StartToStop => track_range.begin,
+                Direction::StopToStart => track_range.end,
+            };
+            extend_path_range_map(
+                &mut res,
+                &mode_range_map,
+                range_entry,
+                offset,
+                track_range.direction,
+            );
+        }
+        offset += track_range.end - track_range.begin;
+    }
+
+    res
+}
+
+/// A struct used for the result of the catenaries_on_path endpoint
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+struct CatenaryRange {
+    begin: f64,
+    end: f64,
+    mode: String,
+}
+
+impl CatenaryRange {
+    fn list_from_range_map(range_map: &RangeMap<Float, String>) -> Vec<CatenaryRange> {
+        range_map
+            .iter()
+            .map(|(range, mode)| CatenaryRange {
+                begin: range.start.0,
+                end: range.end.0,
+                mode: mode.to_string(),
+            })
+            .collect()
+    }
+}
+
+#[derive(Debug, Serialize)]
+#[cfg_attr(test, derive(Deserialize))]
+struct CatenariesOnPathResponse {
+    catenary_ranges: Vec<CatenaryRange>,
+    warnings: Vec<PathfindingError>,
```

Comment:
We can transform the `PathfindingError` into `InternalError` to have these warnings correctly serialized.

```suggestion
    warnings: Vec<InternalError>,
```

Reply:
Done